### ISSUE TITLE
Massive memory spike sometimes on a Rust project

### ISSUE THREAD

**Issue by PgBiel** [zed-industries/zed#8436]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Hi, I sometimes get a very nasty memory leak / spike while I'm working on a Rust project. When it is triggered, memory usage is filled very quickly; Zed can reach 40 GB of memory consumption within a few minutes before requiring a force quit (Zed is completely frozen since the beginning of the leak, with high CPU usage). Not sure what causes it yet; last time it was after selecting a block of code, at least. Worth mentioning that it is a relatively large Rust project.

I've attached some logs from around the moment of the latest leak. I was told to share the memory usage of `rust-analyzer` at the time of the leak. On the latest one, I believe it was Zed using up all the memory, not `rust-analyzer` itself, but I'll try to confirm this (and/or gather further details) if it happens again.

### Environment

Zed: v0.123.6 (Zed)
OS: macOS 13.0.1
Memory: 8 GiB
Architecture: aarch64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

<details>
<summary>Possibly related errors in the logs</summary>

<p>

```
2024-02-26T01:09:46-03:00 [WARN] Generic lsp request to rust-analyzer failed: Invalid Range
[... line above is repeated tens of times, repetitions omitted ...]
2024-02-26T01:09:46-03:00 [WARN] Generic lsp request to rust-analyzer failed: Invalid Range
2024-02-26T01:09:46-03:00 [WARN] Generic lsp request to rust-analyzer failed: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 969}, offset: 188, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 969}, offset: 188, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 969}, offset: 188, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 969}, offset: 188, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 969}, offset: 188, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 969}, offset: 188, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 969}, offset: 188, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:46-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:47-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:48-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:51-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
2024-02-26T01:09:56-03:00 [WARN] request completed with error: request or operation took longer than the configured timeout time
2024-02-26T01:09:56-03:00 [ERROR] crates/client/src/telemetry.rs:583: request or operation took longer than the configured timeout time

Caused by:
    [28] Timeout was reached
2024-02-26T01:09:57-03:00 [ERROR] inlay hint update task for range Anchor { timestamp: Lamport {0: 819}, offset: 0, bias: Right, buffer_id: Some(BufferId(2)) }..Anchor { timestamp: Lamport {0: 727}, offset: 4, bias: Right, buffer_id: Some(BufferId(2)) } failed: inlay hint fetch task: inlay hints LSP request: Invalid Range
```

</p>

</details>

**Comment by mrnugget** [comment]
@PgBiel do you have Xcode and could record a profile of when that happens? It's really hard to get to the bottom of this without circling in on when exactly it happens.

Your logs make it sound like you do have inlay hints enabled when that happens though — is that correct?

@jnathanh I think your issue is "unrelated" to the specifics of this one here, can you open another ticket?

**Comment by PgBiel** [comment]
> @PgBiel do you have Xcode and could record a profile of when that happens? It's really hard to get to the bottom of this without circling in on when exactly it happens.

I don't have Xcode, but I'll keep that in mind next time. Unfortunately, I am now on a Linux PC so I won't be able to provide further info for now (sorry!), but I hope someone with a similar problem manages to bring some more useful info to the table.

For reference, though, I was editing the Typst codebase (at https://github.com/typst/typst).

> Your logs make it sound like you do have inlay hints enabled when that happens though — is that correct?

Yes (they're always enabled).

**Comment by mrnugget** [comment]
> For reference, though, I was editing the Typst codebase (at https://github.com/typst/typst).

Great! That's fantastic to know. Anything specific you did there? Any file you had open that triggered this?

**Comment by jnathanh** [comment]
> @jnathanh I think your issue is "unrelated" to the specifics of this one here, can you open another ticket?

opened a new issue: https://github.com/zed-industries/zed/issues/8680

**Comment by PgBiel** [comment]
> > For reference, though, I was editing the Typst codebase (at https://github.com/typst/typst).
> 
> Great! That's fantastic to know. Anything specific you did there? Any file you had open that triggered this?

I believe I was editing the files under `crates/typst/layout/grid` (on the latest main commit - before, they were in my own branch).

**Comment by JosephTLyons** [comment]
This memory leak has been a long-standing issue that we've tried to fix multiple times - unfortunately, we can't seem to find concrete steps to reproducing it.

**Comment by mocenigo** [comment]
I am having the same problem, which sadly makes zed useless for now (it is very nice and promising though). I load a large file, say, 130Mb of text, and at some point the memory starts ballooning. It usually starts to increase after a save operation. Then in Activity Monitor I quickly see that it is using 2.6Gb, then 4.1, then 8, then 13Gb... It is like some structure is constantly doubling.

**Comment by mocenigo** [comment]
BTW @mrnugget , I have XCode and have built Zed in debug mode. How do I capture a profile?

**Comment by mrnugget** [comment]
@mocenigo thank you! That'd be fantastic if you can reproduce it. _If_ you can reliably reproduce it with a specific file and you can share it, you can also email it to me and I'll try.

But here's a video of how to capture memory allocations for a debug build using `Instruments.app`:



https://github.com/zed-industries/zed/assets/1185253/b91e5cf6-c3f7-433c-a194-3b10c5f15889

**Comment by mrnugget** [comment]
I looked into this today, using the profiles that @mocenigo sent me, along with the `*.tex` file they used.

I can't reproduce the issue without the [zed-latex](https://github.com/rzukic/zed-latex) extension. I don't get any memory leaks. If I install the extension though and open the file, I get massive memory leaks. The profile shows tree-sitter calls, so I suspect that something is off with the tex tree-sitter grammar. It's not the extension itself, since the memory leak even shows up when I comment out everything in the extension and rebuild it locally.

That doesn't explain the problem with memory spikes in Rust problems. It makes me wonder whether the memory spike some of you are seeing isn't a pathological case for some tree-sitter grammars/parsers.

**Comment by mrnugget** [comment]
@PgBiel I checked out the typst codebase again, clicked around, opened a bunch of files, open `crates/typst/src/layout/grid/layout.rs`, made some changes in there — I can't get it to leak memory. Memory consumption might go up, but it also goes down again, quickly.

Can you reliably reproduce it?

**Comment by PgBiel** [comment]
Hey @mrnugget. Unfortunately I don't have a reliable way to reproduce it, and it's been a while since I've last tested it (and even more so experienced it). I'll let you know if I face this problem again, hopefully with more details.

In the meantime, feel free to close the issue if that is appropriate.

**Comment by mocenigo** [comment]
> I looked into this today, using the profiles that @mocenigo sent me, along with the `*.tex` file they used.
> 
> I can't reproduce the issue without the [zed-latex](https://github.com/rzukic/zed-latex) extension. I don't get any memory leaks. If I install the extension though and open the file, I get massive memory leaks. The profile shows tree-sitter calls, so I suspect that something is off with the tex tree-sitter grammar. It's not the extension itself, since the memory leak even shows up when I comment out everything in the extension and rebuild it locally.
> 
> That doesn't explain the problem with memory spikes in Rust problems. It makes me wonder whether the memory spike some of you are seeing isn't a pathological case for some tree-sitter grammars/parsers.

I strongly believe that there is at least one issue here. 

If you open a very large file, and perform some edits, you will see the memory usage spike, as if large structures are copied but never released. I understand for instance the existende of undo buffers. But, if you append a whole copy of the original text to itself, and the text is 130Mb, the memory consumption should not go up by 2Gb. While I understand that you may have GPUI buffers around, you do not need to keep them for UNDOs. This issues is probably independent of file size, but it is easier to observe with large ones.

Then, if you close the window, the memory does NOT go down. It is not released. 

In other words, Zed does not free memory properly.

**Comment by mrnugget** [comment]
@mocenigo take a look at this: https://github.com/zed-industries/zed/pull/10321

I think one half of what you're seeing is what this PR addresses (high memory usage) and the other part might be `MiMalloc`, which is very aggressive in my tests when it comes to allocating memory.

**Comment by mrnugget** [comment]
@LuarRaNa can you create a separate ticket for this with exact steps to reproduce?

**Comment by chibuzoro-salesfloor** [comment]
I open a few rust files, and terminal. Do nothing, wait an hour or two, operating system runs out of memory. Culprit Zed with over 40gb consumed.
Zed version: 0.152.4
MacOS: Sonoma
Aarc: Intel/Amd64

**Comment by mrnugget** [comment]
@chibuzoro-salesfloor is this easy to reproduce with an example project? i.e. does it happen if you open some files in https://github.com/BurntSushi/ripgrep?

**Comment by mymtw** [comment]
I have the same problem like [mrnugget](https://github.com/mrnugget) 
latest zed version
Opened small project on rust.
then moved from pc (smth near 30mins)
then zed app hanged
Idk, my pc: ubuntu 22.04, 32gb ram, ryzen 3600, rtx2060 super

**Comment by mocenigo** [comment]
I wonder whether it isn’t actually tree sitter.Am 28.01.2025 um 14:53 schrieb Igors K ***@***.***>:﻿
I have the same problem like mrnugget
latest zed version
Opened small project on rust.
then moved from pc (smth near 30mins)
then zed app hanged
Idk, my pc: ubuntu 22.04, 32gb ram, ryzen 3600, rtx2060 super

**Comment by maxbrunsfeld** [comment]
We still have not been able to reproduce this. If anyone could find reproducible steps to get the memory usage spike, it would probably make this very easy to fix.

Thanks everyone for providing info.

**Comment by non-descriptive** [comment]
Did exactly that - generated that huge json and opened it. 
On open process manager shows that some IPC process at tmp file starts to grow (fork?).
Displayed command looks like
```
~/.local/zed.app/libexec/zed-editor zed-cli:///tmp/.tmp1cESzw/socket
```
At the same time `package-version-server` at `~/.local/share/zed/languages/` ups and eats about 5 GiBs fro the first time. On the second time spikes to twice that, but quickly drops back to 5 GiB, but IPC socket al the same time do not shrink back and do not release memory when file is closed. If repeat the same thing several times OOM killer comes for you. 
Debian 12, zed 0.172.11 `76db69d7de1a73c7e443e909d7faef17a88fe0a7`.

**Comment by non-descriptive** [comment]
> (duplicate?)

Unlikely. They have some windows issues according to log output. This issue likely related to language server and tokenization, parts, but unclear where to look.

**Comment by notpeter** [comment]
While I appreciate folks commenting here, I believe there are likely multiple memory leaks being discussed -- none of which seem to have an actionable set of steps to reproduce.

For folks running into this, please run a tool like pstree (`brew install pstree` on macos) or use the "Show All Processes Hierarchically" under the Memory tab of Activity monitor so that you can see whether it is Zed itself or a subprocess (rust-analyzer, LSP, etc) which is spiking memory usage. 

> @non-descriptive
> At the same time package-version-server at ~/.local/share/zed/languages/ ups and eats about 5 GiBs fro the first time. On the second time spikes to twice that, but quickly drops back to 5 GiB, but IPC socket al the same time do not shrink back and do not release memory when file is closed. If repeat the same thing several times OOM killer comes for you.
Debian 12, zed 0.172.11 76db69d7de1a73c7e443e909d7faef17a88fe0a7.

This sounds entirely unrelated to this issue -- if you can reliable make [zed-industries/package-version-server](https://github.com/zed-industries/package-version-server) consume 5GB of ram, please open an issue with steps to reproduce.

My apologies that folks are hitting OOM situations, it sucks. But this issue in its current state is not actionable so I'm going to move it to discussions.  Please feel free to open new issues if you have more info about what (zed, a zed extension or a subprocess) is using excess memory and something to trigger that behavior (e.g. an open source project, steps to reproduce), but "after a while, on my closed source project, OOMkiller does it's thing" isn't enough for us to go on, sorry.

Thanks for reporting all.
